synapse/handlers/auth.py

```diff
@@ -1578,13 +1584,20 @@ def generate_access_token(
         return macaroon.serialize()
 
     def generate_short_term_login_token(
-        self, user_id: str, duration_in_ms: int = (2 * 60 * 1000)
+        self,
+        user_id: str,
+        duration_in_ms: int = (2 * 60 * 1000),
+        auth_provider_id: Optional[str] = None,
     ) -> str:
         macaroon = self._generate_base_macaroon(user_id)
         macaroon.add_first_party_caveat("type = login")
         now = self.hs.get_clock().time_msec()
         expiry = now + duration_in_ms
         macaroon.add_first_party_caveat("time < %d" % (expiry,))
+        if auth_provider_id is not None:
+            macaroon.add_first_party_caveat(
+                "auth_provider_id = %s" % (auth_provider_id,)
+            )
```

Comment:
A macaroon that doesn't have an `auth_provider_id` caveat allows clients to add any `auth_provider_id` themselves. I think we want to add a `auth_provider_id = None` (or equivalent) when `auth_provider_id` is None?

Reply:
What is the advantage of storing this information in a macaroon instead of recording it internally in the database?

Reply:
> A macaroon that doesn't have an auth_provider_id caveat allows clients to add any auth_provider_id themselves. I think we want to add a auth_provider_id = None (or equivalent) when auth_provider_id is None?

we were actually setting it to something not-None on all code paths. I've updated the method signature to enforce this.

Reply:
> What is the advantage of storing this information in a macaroon instead of recording it internally in the database?

it's just easier! To store it in the database, we'd have to add a table, do a bunch of sql to insert and select rows, and do a bunch more sql to delete rows once they expire. Not having to do the database queries seems more efficient too.

Reply:
> it's just easier! 

/me looks at the size of this PR :p

Reply:
:-p

This PR is mostly moving existing stuff about. Storing it in the DB would be a bunch of new code.